1832 add tutorial for temporal encoding

p-offtermatt opened the pull request:
<!-- Please ensure that your PR includes the following, as needed -->

- [ ] Tests added for any new code
- [ ] Ran `make fmt-fix` (or had formatting run automatically on all files edited)
- [ ] Documentation added for any new functionality
- [ ] Entries added to [./unreleased/][unreleased] for any new functionality

[unreleased]: https://github.com/informalsystems/apalache/tree/unstable/.unreleased

thpani reviewed (commented):
This is a great writeup!

I think it's all understandable, though I don't know what a person without prior knowledge about model checking algorithms would make of it. Maybe someone who can read LTL but doesn't know the CTL/LTL model checking algorithms should review this as well – @konnov WDYT?

Two major issues that I'd like to discuss:
1. The TLA+ model differs a bit from the prose specification / a real-life traffic light. I think we should at least account for that.
2. The explanation of `_unroll` variables needs a bit of work, I think. AFAIR these are introduced to encode that `[]f` holds in all states of the loop (as opposed to a single state for the subformula auxiliary variable), right? From the description here it's not obvious if this is only relevant for `[]` subformulae.

Otherwise, just some suggestions for wording etc. in the comments.

Also, are you sure you don't want to merge this directly into unstable?

p-offtermatt commented on docs/src/tutorials/temporal-properties.md:
````diff
@@ -0,0 +1,394 @@
+    /\ __loop_requestedGreen = FALSE
+```
+
+Finally, let's discuss `RequestWillBeFulfilled_init`.
+Recall that the variables for subformulas store
+the evaluations of the subformulas at the current point in the execution.
+This means that when we are at the end of the execution, we don't know
+whether the variable for the whole formula, e.g. `☐(requestedGreen ⇒ ♢isGreen)`, was true in the first state.
+That's what `RequestWillBeFulfilled_init` is for:
+it tells us whether in the first state of the execution, our temporal
+property holds true
+(recall that we named the property "RequestWillBeFulfilled").
````
I added more explanation.

thpani replied:
I'm afraid I still don't understand.  This is probably obvious if one knows the entire construction, which I don't.
To clarify, is `☐(requestedGreen ⇒ ♢isGreen)` in the first state a prophecy variable, or is it constrained st we get a model value for it after finding a counter-example / exhaustively exploring the search space?

p-offtermatt replied:
It is a "prohecy variable". If the variable is false, then
in this trace, the formula will not hold. However, only a trace where the prohecy variable says false, and the model checker finds a valid loop (satisfying some constraints wrt the model and aux. variables) is a problem, so the prophecy variable being false does not yet constitute a bad trace. It being false and then later being able to find a valid loop is a counterexample.

thpani replied:
I see, thanks for the explanation!
It seems to me that there is no way of motivating the `_init` variable without knowing the construction. Which would be useful for someone to understand the construction, but this tutorial is aimed at users checking temporal properties. Maybe we should just say that it is an artifact of the model checking algorithm?

thpani replied:
We could give an intuition, e.g., in all states `_init` encodes whether the property holds in the first state, without giving details about why this is necessary.

thpani replied:
I think this is still an issue.

WDYT about my suggestion above?
> Maybe we should just say that it is an artifact of the model checking algorithm?
>  We could give an intuition, e.g., in all states `_init` encodes whether the property holds in the first state, without giving details about why this is necessary.

p-offtermatt replied:
My bad, one commit got lost on the way. I've added a paragraph.

thpani replied:
looks good, thx!

p-offtermatt commented:
Thanks for the great feedback Thomas!
I've incorporated the suggestions and rewritten some things, take a look if you want.
As for merging into unstable: This has a "soft dependency" on the counterexample writer, since otherwise the produced counterexamples look nothing like in the tutorial. So I'll merge this after the counterexample writer is in.

thpani commented on docs/src/tutorials/temporal-properties.md:
````diff
@@ -0,0 +1,394 @@
+    ...
+```
+There are three groups of variables: those that look like formulas, e.g. `☐(requestedGreen ⇒ ♢isGreen)`,
+those that look like formulas and end with `_unroll`,
+e.g., and the variable `RequestWillBeFulfilled_init`.
````
This is not resolved yet.

thpani commented:
@p-offtermatt I think there's still some open issues; I've marked the respective comments unresolved in my original review. Can you please check and address them inline?

thpani commented on docs/src/tutorials/temporal-properties.md:
```diff
@@ -0,0 +1,394 @@
+
+What about the `_unroll` variables? There is one `_unroll` variable for each temporal operator in the formula.
+For example, `☐(requestedGreen ⇒ ♢isGreen)_unroll` is the unroll-variable for the
+leading box operator. These are necessary because to know whether a formula like
+`[]isGreen` holds in the last state of the loop, we need to know whether
+`isGreen` holds in all states of the loop. However, when we are in the last state of the loop,
+we won't see the state of the loop again! So we need to store this information when we traverse the loop.
```
I don't understand what state "we won't see the state of the loop again".

Also, if something is done by an algorithm, I prefer to say it that we instead of using a generic "we".

p-offtermatt replied:
Rewrote this a bit and replaced "we" where appropriate

p-offtermatt commented:
@thpani I just noticed why I thought commits were missing - I pushed to the wrong branch... 😔
Sorry about that - I think my workflow of resolving discussions when I address locally isn't working so well.
My bad, I'll cherry pick the commits in later.

thpani commented:
@p-offtermatt Oh I see, no worries!
Git can be a tricky beast :)

p-offtermatt commented on docs/src/tutorials/temporal-properties.md:
```diff
@@ -0,0 +1,394 @@
+`[]isGreen` holds in the last state of the loop, we need to know whether
+`isGreen` holds in all states of the loop. However, when we are in the last state of the loop,
+we won't see the state of the loop again! So we need to store this information when we traverse the loop.
+That's why we have an extra variable, which essentially tells us whether `isGreen` holds on all states of the loop.
```
I added how this works for `<>` operators

p-offtermatt commented on docs/src/tutorials/temporal-properties.md:
````diff
@@ -0,0 +1,394 @@
+
+Now, we are ready to specify properties we are interested in.
+For example, when green is requested, at some point afterwards the light should actually turn green.
+We can write the property like this:
+
+```
+{{#include TrafficLight.tla:prop}}
+```
````
Added this below

p-offtermatt commented on docs/src/tutorials/temporal-properties.md:
```diff
@@ -0,0 +1,394 @@
+
+The looping trace in this case can be visualized like this:
+
+![A counterexample trace for the property [](requestedGreen => <>isGreen)](img/counterexample.png)
```
Should be there now

p-offtermatt commented on docs/src/tutorials/temporal-properties.md:
```diff
@@ -0,0 +1,415 @@
+Let's first understand how Apalache can identify looping executions using auxiliary variables.
+The auxiliary variable `__loop_InLoop` is true in exactly the states belonging to the loop.
+Additionally, at the first state of the loop, i.e., when `__loop_InLoop` switches from false to true,
+we store the valuation of all variables in an auxiliary copy whose name is prefixed by `__loop_`.
```
```suggestion
we store the valuation of all variables in an auxiliary copy whose name is prefixed by `__loop_`.
```

```suggestion
we store the valuation of all variables in an auxiliary copy whose name is prefixed by `__loop_`.
Before the first state of the loop, the values of the `__loop_` variables is unspecified.
```

p-offtermatt commented on docs/src/tutorials/TrafficLight.tla:
```diff
@@ -0,0 +1,81 @@
+(* requesting green light *)
+\* The switch to green can only be requested when the light is not green, and
+\* the switch has not *already* been requested since the light last turned green.
+RequestGreen_Guard ==
+    /\ ~isGreen
+    /\ ~requestedGreen
```
I explicitly talk about the simplifications in the text now.

konnov commented on docs/src/tutorials/temporal-properties.md:
```diff
@@ -0,0 +1,416 @@
+However, to understand the counterexample, you don't need to go into the technical details of the encoding.
+We'll go explain the counterexample in the following.
+
+## Counterexamples encode traces with a loop
```
The notion of a trace appears out of nowhere. We usually use the word "computation" and the canonical TLA+ notion is that of a behavior. I think you are right in that we have to talk about traces here, as we decorate computation states with additional predicates. It would be good to have the notion of a trace quickly introduced in one paragraph.

p-offtermatt replied:
The nomenclature is from https://apalache.informal.systems/docs//tutorials/symbmc.html?highlight=trace#symbolic-traces, so I'll just link that

konnov commented on docs/src/tutorials/temporal-properties.md:
```diff
@@ -0,0 +1,416 @@
+While some can be very helpful to understand counterexamples,
+many are mostly noise.
+
+Let's first understand how Apalache can identify looping executions using auxiliary variables.
```
Let's use the word "lasso" for executions with a loop, to have the stable vocabulary.

konnov commented on docs/src/tutorials/temporal-properties.md:
```diff
@@ -0,0 +1,416 @@
+Let's first understand how Apalache can identify looping executions using auxiliary variables.
+The auxiliary variable `__loop_InLoop` is true in exactly the states belonging to the loop.
+Additionally, at the first state of the loop, i.e., when `__loop_InLoop` switches from false to true,
+we store the valuation of all variables in an auxiliary copy whose name is prefixed by `__loop_`.
```
```suggestion
we store the valuation of all variables in a shadow copy whose name is prefixed by `__loop_`.
```

konnov commented on docs/src/tutorials/temporal-properties.md:
```diff
@@ -0,0 +1,416 @@
+The auxiliary variable `__loop_InLoop` is true in exactly the states belonging to the loop.
+Additionally, at the first state of the loop, i.e., when `__loop_InLoop` switches from false to true,
+we store the valuation of all variables in an auxiliary copy whose name is prefixed by `__loop_`.
+Before the first state of the loop, the values of the `__loop_` variables are unspecified.
```
```suggestion
Before the first state of the loop, the values of the `__loop_` variables may carry arbitrary values.
```

They are specified in the counterexample.

konnov commented on docs/src/tutorials/temporal-properties.md:
```diff
@@ -0,0 +1,416 @@
+
+In this tutorial, we will show how Apalache can be used to decide temporal
+properties that are more general than invariants.
+This tutorial will be most useful to you if you have a basic understanding of
+linear temporal logic, e.g. the semantics of `<>` and `[]` operators.
+Further, we assume you are familiar with TLA+, but expert knowledge is not necessary.
```
I wanted to give a link to a video lecture by Leslie Lamport on that topic, but it looks like there is no lecture where he is presenting `<>`, or maybe I have not found it. Maybe give a reference to "Specifying Systems"?

thpani replied:
There is a brief explanation of `<>` in video 9a.
But maybe Hillel's page on temporal operators is more informative? https://learntla.com/temporal-logic/operators/

p-offtermatt replied:
Added a link to Hilel's page

konnov commented on docs/src/tutorials/temporal-properties.md:
```diff
@@ -0,0 +1,416 @@
+# Temporal properties and counterexamples
+
+**Difficulty: Red trail – Medium**
+
```
Add yourself as an author.

konnov commented on docs/src/tutorials/temporal-properties.md:
```diff
@@ -0,0 +1,416 @@
+Recall that the temporal property we want to check is `[](requestedGreen => <>isGreen)`.
+That's also the name of one of the variables: The value of the variable 
+`☐(requestedGreen ⇒ ♢isGreen)` tells us whether starting in the current state, the
+formula `[](requestedGreen => <>isGreen)` holds. Since we are looking at a counterexample, it is not
+surprising that the formula does not hold in state 0.
```
```suggestion
formula `[](requestedGreen => <>isGreen)` holds. Since we are looking at a counterexample to this formula, it should not be
surprising that the formula does not hold in the initial state of the counterexample.
```

konnov reviewed (approved):
Great tutorial. I have left a few comments. IMO, we can merge it and then think about further improvements. For instance, we could present the counterexamples as a table (since it is the tableaux encoding) or draw a trace with the values of the predicates below the states. But let's keep it for follow-up PRs.

thpani reviewed (approved):
LGTM, modulo Igor's comments.

> For instance, we could present the counterexamples as a table (since it is the tableaux encoding) or draw a trace with the values of the predicates below the states. But let's keep it for follow-up PRs.

Good ideas! @p-offtermatt, can you make sure we track them in an issue?

p-offtermatt commented:
I've addressed Igor's comments and added an issue for general improvements on the tutorial:
 #1877
